### PR TITLE
Or type fix

### PR DESCRIPTION
**Pull request by harshdoesdev**

### CODE REVIEW

File: ftd/t/js/80-or-type-constant.ftd [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+-- constant integer large: 3
+
+-- end: button-type
+
+
+-- button-type b: small
+
+-- ftd.text: !Smol
+text if { b == button-type.small }: Smol
```
Comment (amitu) [inline ftd/t/js/80-or-type-constant.ftd:15]: Why not this?

```suggestion
text if { b == small }: Smol
```

In line 12 we used just `small`, so we should use it in 15 also.

Comment (harshdoesdev) [reply ftd/t/js/80-or-type-constant.ftd:15]: Okay, got it.

**Comment by amitu** [comment]
Eagerly awaiting this getting done!

File: ftd/t/ast/99-test.ftd [hunk 2]
```diff
@@ -1,6 +0,0 @@
--- ftd.column:
```
Comment (amitu) [inline ftd/t/ast/99-test.ftd]: @harshdoesdev why did you delete this?

File: ftd/t/js/80-or-type-constant.ftd [hunk 3]
```diff
@@ -11,6 +11,15 @@
...
 -- button-type b: small
 
+;; A variable with the same name as a variant of a button type
+;; if we comment this out, the value of "small" will be inferred from the right-hand side or-type
+-- integer small: 2
+
+;; Since the specificity of a variable is higher, the value of the variable will be used
+;; and !smol will be printed
+-- ftd.text: !Smol
+text if { b == small }: Smol
```
Comment (amitu) [inline ftd/t/js/80-or-type-constant.ftd:21]: @harshdoesdev This must fail, the type of `b` is `button-type` and `small` is `integer`. You can not compare different types. Make sure this fails, and then write another test case where `small` is of `button-type` type, and then do the equality check.

Comment (harshdoesdev) [reply ftd/t/js/80-or-type-constant.ftd:21]: @amitu this issue has been resolved and I have tested it locally, now it will throw an error if an integer/any-other-value is provided as the value on either side of equality check.

Comment (harshdoesdev) [reply ftd/t/js/80-or-type-constant.ftd:21]: Sorry, one test is still failing

**Comment by harshdoesdev** [comment]
> @harshdoesdev tests are failing.

It should be resolved now

**Comment by harshdoesdev** [comment]
The build for fastn.com was failing on my system, even when using a
different build of fastn. This was due to network issues. I will test it
again and let you know

File: ftd/t/js/80-or-type-constant.html [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+  }
+}
+global["main"] = main;
+fastn_utils.createNestedObject(global, "foo__button_type", fastn.recordInstance({
+  small: 1,
+  medium: 2,
+  large: 3
+}));
+fastn_utils.createNestedObject(global, "foo__b", 1);
+fastn_utils.createNestedObject(global, "foo__small", 1);
```
Comment (Arpita-Jaiswal) [inline ftd/t/js/80-or-type-constant.html:57]:
```suggestion
fastn_utils.createNestedObject(global, "foo__b", global.foo__button_type.get("small"));
```

Can this be done instead?

Also instead of creating a new variable `global.foo__small`, can we do `global.foo__button_type.get("small")`?

File: ftd/t/js/80-or-type-constant.html [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    (function() {
+        let main = function (parent) {
+  let __fastn_super_package_name__ = __fastn_package_name__;
+  __fastn_package_name__ = "foo";
+  try {
+    let parenti0 = fastn_dom.createKernel(parent, fastn_dom.ElementKind.Text);
+    parenti0.setProperty(fastn_dom.PropertyKind.StringValue, fastn.formula([global.foo__b,
+    global.foo__small], function () {
+      if (function () {
+        return (fastn_utils.getter(global.foo__b) == fastn_utils.getter(global.foo__small));
```
Comment (Arpita-Jaiswal) [inline ftd/t/js/80-or-type-constant.html:35]:
```suggestion
        return (fastn_utils.getter(global.foo__b) == fastn_utils.getter(global.foo__button_type.get("small")));
```

**Review by Arpita-Jaiswal — APPROVED** [review]

**Comment by Arpita-Jaiswal** [comment]
@harshdoesdev though I have added some comments but we can go with this for now. In subsequent PR, can we make the changes I have suggested?

File: ftd/t/js/80-or-type-constant.html (hunk 5 above)
Comment (Arpita-Jaiswal) [reply ftd/t/js/80-or-type-constant.html:35]: I see that you have created a new variable `small` using `-- button-type small: small` and you are using it. So I think then it's fine.

Comment (Arpita-Jaiswal) [reply ftd/t/js/80-or-type-constant.html:35]: But I want another example where you'll not create any variable and use the or-type variant directly
```
if: { b == medium }
```
Where `medium` is not a variable

**Review by Arpita-Jaiswal — APPROVED** [review]

File: ftd/t/js/80-or-type-constant.ftd [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+-- fancy-button: $b
+
+-- component fancy-button:
+caption button-type bt:
+
+-- ftd.text: !Smol
+text if { fancy-button.bt == small }: Smol
```
Comment (amitu) [inline ftd/t/js/80-or-type-constant.ftd:29]: @harshdoesdev in all your examples you have used `small`, does this `small` refer to the variant or the variable? Show example with and without variable both.

File: ftd/t/js/80-or-type-constant.ftd [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+-- button-type b: small
+
+;; A variable with the same name as a variant of a button type
+;; if we comment this out, the value of "small" will be inferred from the right-hand side or-type
+-- button-type small: small
+
+;; Since the specificity of a variable is higher, the value of the variable will be used
+;; and smol will be printed
+-- ftd.text: !Smol
+text if { b == small }: Smol
```
Comment (amitu) [inline ftd/t/js/80-or-type-constant.ftd:21]:
```suggestion
-- ftd.text: !Smol
text if { b == small }: Smol

-- ftd.text: not medium
if: { b != medium }
```

Comment (amitu) [reply ftd/t/js/80-or-type-constant.ftd:21]: @harshdoesdev make test pass after accepting this suggestion.

Comment (harshdoesdev) [reply ftd/t/js/80-or-type-constant.ftd:21]: Ok, got it

Comment (harshdoesdev) [reply ftd/t/js/80-or-type-constant.ftd:21]: @amitu I have updated the tests and the tests are passing

**Review by amitu — APPROVED** [review]
